File: java/src/main/java/nl/craftsmen/workshops/reactivemeetup/exercises/one/Exercise11.java
package nl.craftsmen.workshops.reactivemeetup.exercises.one;

import static nl.craftsmen.workshops.reactivemeetup.util.Utils.*;

import nl.craftsmen.workshops.reactivemeetup.domain.geometry.Shape;
import nl.craftsmen.workshops.reactivemeetup.util.ExampleStreams;
import rx.Observable;

public class Exercise11 {

	public static void main(String[] args) {
		Observable<Shape> shape$ = ExampleStreams.shape$();
		
		// ASSIGNMENT: Use the reduce operator to find the shape that has the largest surface area. Store the result in the
		// largestShape$ stream.
		//
		// HINT: Reduce does not emit intermediate results, it may therefore take some time before the result is available and
		// printed to the console.
		
		Observable<Shape> largestShape$ = unknown(); // ???
		
		// If implemented correctly, only one line is printed to console:
		//   Circle 40: 5026.548245743669
		
		largestShape$.map((i) -> i.getName() + ": " + i.calculateArea())
			.subscribe(System.out::println);

		waitForStreamToComplete(largestShape$);
	}
}

File: javascript/src/exercises/one/exercise-11.js
const Rx = require('rxjs/Rx');
const shape$ = require('../../util/example-streams.js').shape$;

// ASSIGNMENT: Use the reduce operator to find the shape that has the largest surface area.
//
// NOTE: RxJS has a max operator (not discussed in the presentation), which you normally would use for scenario's like this assignment.
// However, for this exercise we want you use the reduce operator to become more familiar with that operator.
//
// HINT: Reduce does not emit intermediate results, it may therefore take some time before the result is available and printed to the
// console.

const largestShape$ =shape$
	.reduce((a, b) => {
		if (a.calculateArea() > b.calculateArea()) {
			return a;
		}
		return b;
	});


// If implemented correctly, only one line is printed to console:
//   Circle 40: 5026.548245743669

largestShape$.map((i) => i.getName() + ': ' + i.calculateArea())
	.subscribe((i) => console.log(i));